api/src/main/scala/uk/ac/wellcome/platform/api/finatra/exceptions/ElasticsearchResponseExceptionMapper.scala

```diff
@@ -0,0 +1,111 @@
+package uk.ac.wellcome.platform.api.finatra.exceptions
+
+import javax.inject.{Inject, Singleton}
+
+import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
+import com.twitter.finagle.http.{Request, Response}
+import com.twitter.finatra.http.exceptions.ExceptionMapper
+import com.twitter.finatra.http.response.ResponseBuilder
+import com.twitter.inject.annotations.Flag
+import com.twitter.inject.Logging
+import org.elasticsearch.client.ResponseException
+
+import uk.ac.wellcome.models.Error
+import uk.ac.wellcome.platform.api.models.DisplayError
+import uk.ac.wellcome.platform.api.responses.ResultResponse
+
+@Singleton
+class ElasticsearchResponseExceptionMapper @Inject()(
+  response: ResponseBuilder,
+  @Flag("api.context") apiContext: String,
+  @Flag("api.host") apiHost: String,
+  @Flag("api.scheme") apiScheme: String)
+    extends ExceptionMapper[ResponseException]
+    with Logging {
+
+  val contextUri: String = s"${apiScheme}://${apiHost}${apiContext}"
+
+  private def userError(message: String): DisplayError = {
+    error(s"Sending HTTP 400 for $message")
+    DisplayError(Error(variant = "http-400", description = Some(message)))
+  }
+
+  private def serverError(message: String): DisplayError = {
+    error(s"Sending HTTP 500 for $message")
+    DisplayError(Error(variant = "http-500", description = None))
+  }
+
+  // One possible error is of the form:
+  //
+  //     Result window is too large, from + size must be less than or equal
+  //     to: [10000] but was [100000]. See the scroll api for a more
+  //     efficient way to request large data sets. This limit can be set by
+  //     changing the [index.max_result_window] index level setting.
+  //
+  // In this case, we return a 400 Bad Request error, but we rephrase it
+  // to be shorter and to avoid talking about Elasticsearch concepts.
+  val resultSizePattern =
+    """Result window is too large, from \+ size must be less than or equal to: \[([0-9]+)\]""".r.unanchored
```

Comment:
Sorry for being a bit late in reviewing this, but I wonder if it's worth trying doing this without matching on the message. It seems a bit brittle to me: we might upgrade the library and find out that this is broken because the message has changed slightly.

Reply:
I’m not opposed to the idea, but I couldn’t see anything more specific to match on in the JSON:

```json
{
  "error": {
    "root_cause": [
      {
        "type": "query_phase_execution_exception",
        "reason": "Result window is too large, from + size must be less than or equal to: [10000] but was [100010]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting."
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": ".kibana",
        "node": "nrhHcUIZQxiEASp_g93UaQ",
        "reason": {
          "type": "query_phase_execution_exception",
          "reason": "Result window is too large, from + size must be less than or equal to: [10000] but was [100010]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting."
        }
      }
    ]
  },
  "status": 500
}
```

Also, if the message changes, we have a test for this code path, so it wouldn’t fail silently.

Reply:
Yeah, I know, but it still feels wrong to have code that works on what is essentially a logging message. It feels a bit of a hack. Although I don't really have an answer as to how to do it, so fine to leave as is. I'll try poking around a bit